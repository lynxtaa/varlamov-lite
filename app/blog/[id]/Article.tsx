import cn from 'classnames'
import { formatDistanceToNow } from 'date-fns'
import { ru } from 'date-fns/locale'
import { Clock } from 'react-feather'

import ArticleBody from '../../../components/ArticleBody'
import Icon from '../../../components/Icon'
import Link from '../../../components/Link'
import Page from '../../../components/layouts/Page'
import { formatDate } from '../../../lib/formatDate'
import { type ArticleFull } from '../../../lib/varlamovClient'

export type Props = ArticleFull

export default function Article({
	published_at,
	tags,
	title,
	text,
	readingTime,
	topics,
}: Props) {
	return (
		<Page>
			{topics.length > 0 && (
				<div className="flex flex-wrap">
					{topics.map(({ name }, i) => (
						<div
							key={name}
							className={cn(
								'py-1 px-3 border rounded-md border-gray-700 border-solid mb-3 cursor-default bg-gray-900 text-gray-50 dark:bg-gray-200 dark:text-gray-900 font-semibold',
								i < tags.length && 'mr-3',
							)}
						>
							{name}
						</div>
					))}
				</div>
			)}
			{published_at !== null && (
				<time className="flex mb-1">
					{formatDate(published_at)}
					<span className="opacity-70 inline-flex items-center ml-4">
						<Icon icon={<Clock />} className="mr-2 w-4 h-4" />
						{formatDistanceToNow(Date.now() + readingTime, { locale: ru })}
					</span>
				</time>
			)}
			<h1 className="font-bold text-3xl mb-5">{title}</h1>
			<ArticleBody text={text} className="mb-4" />
			<div className="flex flex-wrap">
				{tags
					.map(tag => `#${tag}`)
					.map((tag, i) => (
						<Link
							key={tag}
							href={`/search/${encodeURIComponent(tag)}`}
							className={cn(
								'text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline',
								i < tags.length && 'mr-3',
							)}
						>
							{tag}
						</Link>
					))}
			</div>
		</Page>
	)
}
