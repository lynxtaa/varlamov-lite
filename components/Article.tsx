import Head from 'next/head'
import Image from 'next/image'
import { useMemo, Fragment } from 'react'
import { Home } from 'react-feather'

import styles from './Article.module.css'
import Link from './Link'

type Props = {
	title: string
	text: string
}

export default function Article({ title, text }: Props) {
	const textWithImages: JSX.Element[] = useMemo(() => {
		const pattern = /(!\[.*\]\(http.+\))/g

		return text.split(pattern).map((part, index) => {
			const linkPattern = /!\[(?<width>\d+)x(?<height>\d+)\]\((?<src>http.+)\)/
			const match = part.match(linkPattern)

			if (match?.groups) {
				return (
					<Image
						key={index}
						src={match.groups.src}
						alt=""
						width={match.groups.width}
						height={match.groups.height}
						className={styles.Image}
					/>
				)
			}

			return <Fragment key={index}>{part}</Fragment>
		})
	}, [text])

	return (
		<article className={styles.Article}>
			<div className={styles.backArrow}>
				<Link href="/">
					<Home />
				</Link>
			</div>
			<Head>
				<title>{title}</title>
			</Head>
			<h1>{title}</h1>
			<div className={styles.text}>{textWithImages}</div>
		</article>
	)
}
