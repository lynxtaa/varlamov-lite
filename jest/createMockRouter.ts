import { NextRouter } from 'next/router'

export function createMockRouter(options?: Partial<NextRouter>): NextRouter {
	return {
		basePath: '/',
		pathname: '/',
		route: '/',
		query: {},
		asPath: '/',
		push: jest.fn(() => Promise.resolve(true)),
		replace: jest.fn(() => Promise.resolve(true)),
		reload: jest.fn(),
		prefetch: jest.fn(() => Promise.resolve()),
		back: jest.fn(),
		forward: jest.fn(),
		beforePopState: jest.fn(),
		isFallback: false,
		isLocaleDomain: true,
		events: {
			on: jest.fn(),
			off: jest.fn(),
			emit: jest.fn(),
		},
		isReady: true,
		isPreview: false,
		...options,
	}
}
