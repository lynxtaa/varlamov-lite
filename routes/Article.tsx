import cn from 'classnames'
import { formatDistanceToNow } from 'date-fns'
import { ru } from 'date-fns/locale'
import { useRouter } from 'next/router'
import { Clock } from 'react-feather'

import ArticleBody from '../components/ArticleBody'
import Icon from '../components/Icon'
import Link from '../components/Link'
import Spinner from '../components/Spinner'
import Page from '../layouts/Page'
import { formatDate } from '../lib/formatDate'
import { ArticleFull } from '../lib/varlamovClient'

export type Props = ArticleFull

export default function Article({
	created_at,
	excerpt,
	tags,
	title,
	text,
	previewImageUrl,
	readingTime,
}: Props) {
	const router = useRouter()

	if (router.isFallback) {
		return <Spinner />
	}

	return (
		<Page title={title} description={excerpt} ogImage={previewImageUrl || undefined}>
			{created_at && (
				<time className="flex mb-1">
					{formatDate(created_at)}
					<span className="opacity-70 inline-flex items-center ml-4">
						<Icon icon={<Clock />} className="mr-2 w-4 h-4" />
						{formatDistanceToNow(Date.now() + readingTime, { locale: ru })}
					</span>
				</time>
			)}
			<h1 className="font-bold text-3xl mb-5">{title}</h1>
			<ArticleBody text={text} className="mb-4" />
			<div className="flex flex-wrap">
				{tags.map((tag, i) => (
					<Link
						key={tag}
						href={`/tag/${tag}`}
						className={cn(
							'text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline',
							i < tags.length && 'mr-3',
						)}
					>
						{tag}
					</Link>
				))}
			</div>
		</Page>
	)
}
