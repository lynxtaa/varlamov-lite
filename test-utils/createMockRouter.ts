import { NextRouter } from 'next/router'

export function createMockRouter(options?: Partial<NextRouter>): NextRouter {
	return {
		basePath: '/',
		pathname: '/',
		route: '/',
		query: {},
		asPath: '/',
		push: vi.fn(() => Promise.resolve(true)),
		replace: vi.fn(() => Promise.resolve(true)),
		reload: vi.fn(),
		prefetch: vi.fn(() => Promise.resolve()),
		back: vi.fn(),
		forward: vi.fn(),
		beforePopState: vi.fn(),
		isFallback: false,
		isLocaleDomain: true,
		events: {
			on: vi.fn(),
			off: vi.fn(),
			emit: vi.fn(),
		},
		isReady: true,
		isPreview: false,
		...options,
	}
}
