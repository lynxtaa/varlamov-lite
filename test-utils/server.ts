import { setupServer } from 'msw/node'

const server = setupServer()

export { rest } from 'msw'
export { server }
