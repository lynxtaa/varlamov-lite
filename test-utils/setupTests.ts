import '@testing-library/jest-dom/extend-expect'

import { server } from './server'

vi.mock('next/router', () => ({
	useRouter: vi.fn(),
}))

beforeAll(() =>
	server.listen({
		onUnhandledRequest: 'warn',
	}),
)

afterEach(() => {
	sessionStorage.clear()
	localStorage.clear()

	vi.useRealTimers()

	server.resetHandlers()
})

afterAll(() => server.close())

global.IntersectionObserver = class IntersectionObserver {
	observe = vi.fn()
	root = null
	rootMargin = ''
	disconnect = vi.fn()
	thresholds = []
	takeRecords = vi.fn()
	unobserve = vi.fn()
}
