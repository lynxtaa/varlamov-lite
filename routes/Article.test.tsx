import { render } from '@testing-library/react'
import { NextRouter, useRouter } from 'next/router'
import { Mock } from 'vitest'

import { createMockRouter } from '../test-utils/createMockRouter'

import Article from './Article'

const useRouterMock = useRouter as Mock<[], NextRouter>

beforeEach(() => {
	useRouterMock.mockReturnValue(createMockRouter())
})

it('shows article', () => {
	const now = new Date(2020, 2, 20, 0, 0)
	vi.setSystemTime(now)

	const text = `
		Lorem ipsum dolor sit amet, consectetur adipisicing elit.
		Cumque voluptatum numquam, omnis saepe facere cupiditate
		sit accusamus nesciunt nam debitis voluptatem exercitationem
		fugiat modi, aliquid nostrum vitae! Repellendus atque aliquid
		corrupti animi deleniti voluptatem provident libero facilis
		veritatis aut totam, distinctio saepe reiciendis neque
		ducimus sit eligendi ut architecto iste commodi pariatur
		non soluta ea praesentium? Quas iusto placeat voluptate
		excepturi quia perspiciatis sapiente quae reprehenderit
		ratione nisi iure laudantium asperiores ad
	`.trim()

	const container = render(
		<Article
			id={1}
			uri="1.html"
			excerpt={`${text.slice(0, 100)}...`}
			previewImageUrl="https://varlamov.me/2021/izmajlovo/02.jpg"
			title="Lorem ipsum, dolor sit amet consectetur adipisicing."
			tags={['Москва', 'Россия']}
			text={`
					${text}
					<br>
					<br>
					<img src="/2021/izmajlovo/02.jpg" width="960" height="720">
					<br>
					<i>
					  <span>Фото: <a href="https://www.facebook.com/tatyana.tsarenko.94/posts/1955993087875425">Татьяна Царенко / Facebook</a></span>
					</i>
				`}
			published_at="2021-03-29T18:37:00.000Z"
			readingTime={181800}
			topics={[]}
		/>,
	)

	expect(container).toMatchSnapshot()
})
