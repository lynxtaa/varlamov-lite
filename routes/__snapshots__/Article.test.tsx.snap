// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows article 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="my-0 mx-auto p-4 max-w-3xl"
      >
        <header
          class="flex flex-col mt-4 mb-8"
        >
          <div
            class="flex items-center"
          >
            <h1
              class="text-2xl mr-3 font-bold"
            >
              Блог Ильи Варламова
            </h1>
            <button
              class="ml-auto"
              title="Поиск"
              type="button"
            >
              <div
                class="w-7 h-7"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  style="width: 100%; height: 100%;"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="11"
                    cy="11"
                    r="8"
                  />
                  <line
                    x1="21"
                    x2="16.65"
                    y1="21"
                    y2="16.65"
                  />
                </svg>
              </div>
            </button>
            <button
              class="ml-4"
              title="Переключить тему"
              type="button"
            >
              <div
                class="w-7 h-7"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  style="width: 100%; height: 100%;"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
                  />
                </svg>
              </div>
            </button>
          </div>
        </header>
        <time
          class="flex mb-1"
        >
          29 марта 2021
          <span
            class="opacity-70 inline-flex items-center ml-4"
          >
            <div
              class="mr-2 w-4 h-4"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="width: 100%; height: 100%;"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <polyline
                  points="12 6 12 12 16 14"
                />
              </svg>
            </div>
            3 минуты
          </span>
        </time>
        <h1
          class="font-bold text-3xl mb-5"
        >
          Lorem ipsum, dolor sit amet consectetur adipisicing.
        </h1>
        <div
          class="mb-4"
        >
          
					Lorem ipsum dolor sit amet, consectetur adipisicing elit.
		Cumque voluptatum numquam, omnis saepe facere cupiditate
		sit accusamus nesciunt nam debitis voluptatem exercitationem
		fugiat modi, aliquid nostrum vitae! Repellendus atque aliquid
		corrupti animi deleniti voluptatem provident libero facilis
		veritatis aut totam, distinctio saepe reiciendis neque
		ducimus sit eligendi ut architecto iste commodi pariatur
		non soluta ea praesentium? Quas iusto placeat voluptate
		excepturi quia perspiciatis sapiente quae reprehenderit
		ratione nisi iure laudantium asperiores ad
					
          <br />
          
					
          <br />
          
					
          <img
            alt=""
            data-nimg="1"
            decoding="async"
            height="720"
            loading="lazy"
            src="/_next/image?url=%2F2021%2Fizmajlovo%2F02.jpg&w=1920&q=75"
            srcset="/_next/image?url=%2F2021%2Fizmajlovo%2F02.jpg&w=1080&q=75 1x, /_next/image?url=%2F2021%2Fizmajlovo%2F02.jpg&w=1920&q=75 2x"
            style="color: transparent;"
            width="960"
          />
          
					
          <br />
          
					
          <i>
            
					  
            <span>
              Фото: 
              <a
                class="group hover:underline underline"
                href="https://www.facebook.com/tatyana.tsarenko.94/posts/1955993087875425"
                rel="noopener noreferrer"
                target="_blank"
              >
                Татьяна Царенко / Facebook
                <div
                  class="inline-block ml-1 w-3 h-3 opacity-70 group-hover:opacity-100 transition-opacity"
                >
                  <svg
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    style="width: 100%; height: 100%;"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
                    />
                    <polyline
                      points="15 3 21 3 21 9"
                    />
                    <line
                      x1="10"
                      x2="21"
                      y1="14"
                      y2="3"
                    />
                  </svg>
                </div>
              </a>
            </span>
            
					
          </i>
          
				
        </div>
        <div
          class="flex flex-wrap"
        >
          <a
            class="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline mr-3"
            href="/search/%23%D0%9C%D0%BE%D1%81%D0%BA%D0%B2%D0%B0"
          >
            #Москва
          </a>
          <a
            class="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline mr-3"
            href="/search/%23%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D1%8F"
          >
            #Россия
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="my-0 mx-auto p-4 max-w-3xl"
    >
      <header
        class="flex flex-col mt-4 mb-8"
      >
        <div
          class="flex items-center"
        >
          <h1
            class="text-2xl mr-3 font-bold"
          >
            Блог Ильи Варламова
          </h1>
          <button
            class="ml-auto"
            title="Поиск"
            type="button"
          >
            <div
              class="w-7 h-7"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="width: 100%; height: 100%;"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="11"
                  cy="11"
                  r="8"
                />
                <line
                  x1="21"
                  x2="16.65"
                  y1="21"
                  y2="16.65"
                />
              </svg>
            </div>
          </button>
          <button
            class="ml-4"
            title="Переключить тему"
            type="button"
          >
            <div
              class="w-7 h-7"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="width: 100%; height: 100%;"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
                />
              </svg>
            </div>
          </button>
        </div>
      </header>
      <time
        class="flex mb-1"
      >
        29 марта 2021
        <span
          class="opacity-70 inline-flex items-center ml-4"
        >
          <div
            class="mr-2 w-4 h-4"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="width: 100%; height: 100%;"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <polyline
                points="12 6 12 12 16 14"
              />
            </svg>
          </div>
          3 минуты
        </span>
      </time>
      <h1
        class="font-bold text-3xl mb-5"
      >
        Lorem ipsum, dolor sit amet consectetur adipisicing.
      </h1>
      <div
        class="mb-4"
      >
        
					Lorem ipsum dolor sit amet, consectetur adipisicing elit.
		Cumque voluptatum numquam, omnis saepe facere cupiditate
		sit accusamus nesciunt nam debitis voluptatem exercitationem
		fugiat modi, aliquid nostrum vitae! Repellendus atque aliquid
		corrupti animi deleniti voluptatem provident libero facilis
		veritatis aut totam, distinctio saepe reiciendis neque
		ducimus sit eligendi ut architecto iste commodi pariatur
		non soluta ea praesentium? Quas iusto placeat voluptate
		excepturi quia perspiciatis sapiente quae reprehenderit
		ratione nisi iure laudantium asperiores ad
					
        <br />
        
					
        <br />
        
					
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="720"
          loading="lazy"
          src="/_next/image?url=%2F2021%2Fizmajlovo%2F02.jpg&w=1920&q=75"
          srcset="/_next/image?url=%2F2021%2Fizmajlovo%2F02.jpg&w=1080&q=75 1x, /_next/image?url=%2F2021%2Fizmajlovo%2F02.jpg&w=1920&q=75 2x"
          style="color: transparent;"
          width="960"
        />
        
					
        <br />
        
					
        <i>
          
					  
          <span>
            Фото: 
            <a
              class="group hover:underline underline"
              href="https://www.facebook.com/tatyana.tsarenko.94/posts/1955993087875425"
              rel="noopener noreferrer"
              target="_blank"
            >
              Татьяна Царенко / Facebook
              <div
                class="inline-block ml-1 w-3 h-3 opacity-70 group-hover:opacity-100 transition-opacity"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  style="width: 100%; height: 100%;"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
                  />
                  <polyline
                    points="15 3 21 3 21 9"
                  />
                  <line
                    x1="10"
                    x2="21"
                    y1="14"
                    y2="3"
                  />
                </svg>
              </div>
            </a>
          </span>
          
					
        </i>
        
				
      </div>
      <div
        class="flex flex-wrap"
      >
        <a
          class="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline mr-3"
          href="/search/%23%D0%9C%D0%BE%D1%81%D0%BA%D0%B2%D0%B0"
        >
          #Москва
        </a>
        <a
          class="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline mr-3"
          href="/search/%23%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D1%8F"
        >
          #Россия
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
